pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -0,0 +1,210 @@
+// Note: Symmetry is not applicable. We only weigh how incomingPod matches existingPod.
+// Whether existingPod matches incomingPod doesn't contribute to the final score.
+// This is different with the Affinity API.
+func CalculateEvenPodsSpreadPriority(pod *v1.Pod, nodeNameToInfo map[string]*schedulernodeinfo.NodeInfo, nodes []*v1.Node) (schedulerapi.HostPriorityList, error) {
+	result := make(schedulerapi.HostPriorityList, len(nodes))
+	// return if incoming pod doesn't have soft topology spread constraints.
+	constraints := getSoftTopologySpreadConstraints(pod)
+	if len(constraints) == 0 {
+		return result, nil
+	}
+
+	t := newTopologySpreadConstraintsMap(len(nodes))
+	t.initialize(pod, nodes)
+
+	allNodeNames := make([]string, 0, len(nodeNameToInfo))
+	for name := range nodeNameToInfo {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	errCh := schedutil.NewErrorChannel()
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
+		nodeInfo := nodeNameToInfo[allNodeNames[i]]
+		node := nodeInfo.Node()
+		if node == nil {
+			return
+		}
+		// (1) `node` should satisfy incoming pod's NodeSelector/NodeAffinity
+		// (2) All topologyKeys need to be present in `node`
+		if !predicates.PodMatchesNodeSelectorAndAffinityTerms(pod, node) ||
+			!predicates.NodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
+			return
+		}
+		// It's enough to use topologyKey as the "key" of the map.
+		matchCount := make(map[string]int64)
+		for _, existingPod := range nodeInfo.Pods() {
+			podLabelSet := labels.Set(existingPod.Labels)
+			// Matching on constraints is calculated independently.
+			for _, constraint := range constraints {
+				match, err := predicates.PodMatchesSpreadConstraint(podLabelSet, constraint)
+				if err != nil {
+					errCh.SendErrorWithCancel(err, cancel)
+					return
+				}
+				if match {
+					matchCount[constraint.TopologyKey]++
+				}
+			}
+		}
+		// Keys in t.podCounts have been ensured to contain "filtered" nodes only.
+		for _, constraint := range constraints {
+			tpKey := constraint.TopologyKey
+			pair := topologyPair{key: tpKey, value: node.Labels[tpKey]}
+			// For each <pair>, all matched nodes get the credit of summed matchCount.
+			// And we add matchCount to <t.total> to reverse the final score later.
+			for _, nodeName := range t.topologyPairToNodeNames[pair] {
+				atomic.AddInt64(t.podCounts[nodeName], matchCount[tpKey])
+				atomic.AddInt64(&t.total, matchCount[tpKey])
+			}
+		}
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
+	if err := errCh.ReceiveError(); err != nil {
+		return nil, err
+	}
+
+	var maxCount, minCount int64
+	for _, node := range nodes {
+		if t.podCounts[node.Name] == nil {
+			continue
+		}
+		// reverse
+		count := t.total - *t.podCounts[node.Name]
+		if count > maxCount {
+			maxCount = count
+		} else if count < minCount {
+			minCount = count
+		}
+		t.podCounts[node.Name] = &count
```

Comment:
If all nodes end up with the same count, they seem to get 0 priority. Is that ok?

Reply:
Nope. They will be all get 10. I added a testcase covering this.